subsys/net/lib/lwm2m/lwm2m_message_handling.c

```diff
@@ -325,6 +325,9 @@ STATIC int build_msg_block_for_send(struct lwm2m_message *msg, uint16_t block_nu
 		}
 	}
 
+	/* Copy old msg reference to keep send_status_cb */
+	msg->reply->user_data = msg->out.user_data;
```

Comment:
After looking closer into this, it seems that `msg->reply->user_data` is lost, when consecutive block is being prepared to send, here: https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/net/lib/lwm2m/lwm2m_message_handling.c#L2730

The previous `struct coap_reply` is freed, and a new one is allocated. Therefore, instead of involving seemingly unrelated `user_data` field from the output context, wouldn't it make sense to simply store `msg->reply->user_data` locally there, before the next block is prepered, and set it back right after?

Reply:
It would be better than using another user_data pointer as temporary buffer